SeaPkg/Library/BasePeCoffValidationLib/Validation.c

```diff
@@ -0,0 +1,325 @@
+  @return EFI_SECURITY_VIOLATION  The specified buffer in the target image does not match the reference data.
+**/
+EFI_STATUS
+PeCoffImageValidationContent (
+  IN VOID                                *TargetImage,
+  IN IMAGE_VALIDATION_ENTRY_HEADER       *Hdr,
+  IN CONST IMAGE_VALIDATION_DATA_HEADER  *ImageValidationHdr
+  )
+{
+  IMAGE_VALIDATION_CONTENT  *ContentHdr;
+  EFI_STATUS                Status;
+
+  ContentHdr = (IMAGE_VALIDATION_CONTENT *)Hdr;
+  // Ensure "Content" in the header (TargetContent) does not overflow the Auxiliary file buffer.
+  if ((UINT8 *)ContentHdr + sizeof (*ContentHdr) + Hdr->Size > (UINT8 *)ImageValidationHdr + ImageValidationHdr->Size) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: Current entry range 0x%p: 0x%x exceeds reference data limit 0x%x\n",
+      __func__,
+      ContentHdr,
+      sizeof (*ContentHdr) + Hdr->Size,
+      (UINT8 *)ImageValidationHdr + ImageValidationHdr->Size
+      ));
+    Status = EFI_COMPROMISED_DATA;
+    goto Done;
+  }
+
+  if (CompareMem ((UINT8 *)TargetImage + Hdr->Offset, ContentHdr->TargetContent, Hdr->Size) != 0) {
+    DEBUG ((DEBUG_ERROR, "%a: Current entry range 0x%p: 0x%x does not match input content at 0x%p\n", __func__, ContentHdr, Hdr->Size, ContentHdr->TargetContent));
+    Status = EFI_SECURITY_VIOLATION;
+    goto Done;
+  }
+
+  Status = EFI_SUCCESS;
+
+Done:
+  return Status;
+}
+
+/**
+  Validates a specific region in the target image buffer denoted by [Hdr->Offset: Hdr->Offset + Hdr->Size]
+  belongs to a user page that is mapped inside MM and teh page attributes match the requirements specified
+  by the validation entry.
+    @param[in] TargetImage    The pointer to the target image buffer.
+    @param[in] Hdr            The header of the validation entry.
+    @param[in] PageTableBase  The base address of the page table.
+
+    @returns EFI_SUCCESS             The target image passes the validation.
+    @returns EFI_INVALID_PARAMETER   The validation entry has invalid must have and must not have attributes.
+    @returns EFI_INVALID_PARAMETER   The validation entry data size is invalid. It must be a pointer size.
+    @returns EFI_SECURITY_VIOLATION  The target image does not meet the memory attribute requirements.
+**/
+EFI_STATUS
+PeCoffImageValidationMemAttr (
+  IN VOID                           *TargetImage,
+  IN IMAGE_VALIDATION_ENTRY_HEADER  *Hdr,
+  IN EFI_PHYSICAL_ADDRESS           PageTableBase
+  )
+{
+  UINT64                     MemAttr;
+  IMAGE_VALIDATION_MEM_ATTR  *MemAttrHdr;
+  EFI_PHYSICAL_ADDRESS       AddrInTarget;
+  EFI_STATUS                 Status;
+
+  MemAttrHdr = (IMAGE_VALIDATION_MEM_ATTR *)Hdr;
+  if ((MemAttrHdr->TargetMemoryAttributeMustHave == 0) && (MemAttrHdr->TargetMemoryAttributeMustNotHave == 0)) {
+    DEBUG ((DEBUG_ERROR, "%a: Current entry 0x%p has invalid must have 0x%x and must not have 0x%x\n", __func__, MemAttrHdr, MemAttrHdr->TargetMemoryAttributeMustHave, MemAttrHdr->TargetMemoryAttributeMustNotHave));

```

Comment:
Since this is only printed if both are zero, can this be simpler and more helpful?

```suggestion
    DEBUG ((DEBUG_ERROR, "%a: Entry at 0x%p cannot have zero for must and must not have attribute values.\n", __func__, MemAttrHdr, MemAttrHdr->TargetMemoryAttributeMustHave, MemAttrHdr->TargetMemoryAttributeMustNotHave));
```

Reply:
Makes sense to me. Updated!